minorminer/utils/zephyr/node_edge.py

```diff
@@ -0,0 +1,796 @@
+            raise ValueError(
+                f"Expected {self, other} to have the same shape, got {self._shape, other._shape}"
+            )
+        return self._ccoord != other._ccoord
+
+    def __gt__(self, other: ZNode) -> bool:
+        if not isinstance(other, ZNode):
+            raise TypeError(f"Expected {other} to be {type(self).__name__}, got {type(other)}")
+        if self._shape != other._shape:
+            raise ValueError(
+                f"Expected {self, other} to have the same shape, got {self._shape, other._shape}"
+            )
+        return self._ccoord > other._ccoord
+
+    def __ge__(self, other: ZNode) -> bool:
+        if not isinstance(other, ZNode):
+            raise TypeError(f"Expected {other} to be {type(self).__name__}, got {type(other)}")
+        if self._shape != other._shape:
+            raise ValueError(
+                f"Expected {self, other} to have the same shape, got {self._shape, other._shape}"
+            )
+        return self._ccoord >= other._ccoord
+
+    def __lt__(self, other: ZNode) -> bool:
+        if not isinstance(other, ZNode):
+            raise TypeError(f"Expected {other} to be {type(self).__name__}, got {type(other)}")
+        if self._shape != other._shape:
+            raise ValueError(
+                f"Expected {self, other} to have the same shape, got {self._shape, other._shape}"
+            )
+        return self._ccoord < other._ccoord
+
+    def __le__(self, other: ZNode) -> bool:
+        if not isinstance(other, ZNode):
+            raise TypeError(f"Expected {other} to be {type(self).__name__}, got {type(other)}")
+        if self._shape != other._shape:
+            raise ValueError(
+                f"Expected {self, other} to have the same shape, got {self._shape, other._shape}"
+            )
+        return self._ccoord <= other._ccoord
+
+    def __add__(
+        self,
+        shift: PlaneShift | tuple[int],
+    ) -> ZNode:
+        if not isinstance(shift, PlaneShift):
+            shift = PlaneShift(*shift)
+        x, y, k = self._ccoord
+        new_x = x + shift[0]
+        new_y = y + shift[1]
+
+        return ZNode(coord=CartesianCoord(x=new_x, y=new_y, k=k), shape=self._shape)
+
+    def __sub__(
+        self,
+        other: ZNode,
+    ) -> PlaneShift:
+        if not isinstance(other, ZNode):
+            raise TypeError(f"Expected {other} to be {type(self).__name__}, got {type(other)}")
+        if self._shape != other._shape:
+            raise ValueError(
+                f"Expected {self, other} to have the same shape, got {self._shape, other._shape}"
+            )
+        x_shift: int = self._ccoord.x - other._ccoord.x
+        y_shift: int = self._ccoord.y - other._ccoord.y
+        try:
+            return PlaneShift(x_shift=x_shift, y_shift=y_shift)
+        except:
+            raise ValueError(f"{other} cannot be subtracted from {self}")
+
+    def __hash__(self) -> int:
+        return (self._ccoord, self._shape).__hash__()
```

Comment:
Not wrong per se, but it's usually better to call the hash function rather than the dunder method.
```suggestion
        return hash((self._ccoord, self._shape))
```